share/wake/lib/core/json.wake

```diff
@@ -154,18 +154,155 @@ export def x // y =
+
+# Target the minimum JSON language described by the specifications.
+export def normalizeJSONBasic: JSONNormalize =
+    def nfkcKeys =
+        map (_ | editPairFirst unicodeIdentifier)
+    normalizeJSONDefault
+    | setJSONNormalizeString (Pass _.unicodeCanonical)
+    | setJSONNormalizeDouble filterNonDigitJDouble
+    | setJSONNormalizeObject (deduplicateJObjectKeys _.nfkcKeys)
+
+# Fail on any `Double` values which can't be represented numerically.
+# Both JSON specifications describe their double values as allowing decimal or
+# exponential forms, but don't implement the full IEEE standard.
+def filterNonDigitJDouble (n: Double): Result Double Error = match n.dclass
+    DoubleInfinite =
+        failWithError "For compatibility, JSON doubles should not be infinite."
+    DoubleNaN =
+        failWithError "For compatibility, JSON doubles should not be NaN values."
+    _ =
+        Pass n
+
+# Attempt to simplify any `JObject`s which contain multiple instances of a key.
+# While both specifications explicitly allow such duplicate keys, neither
+# requires implementations to do so or describes the behaviour by which they
+# should be handled; therefore, implementations differ in how they handle
+# duplicate keys -- if they do so at all -- and a generator aiming for
+# compatibility shouldn't output objects with duplicate keys.
+#
+# In particular, this function will pass if all instances of the same key also
+# have the same value, or otherwise will combine the values of `JArray` or
+# `JObject` values if all instances of the key have the same type.  It fails if
+# a key is associated with multiple values of different types or different
+# values of a type which cannot be combined.
+def deduplicateJObjectKeys (dict: List (Pair String JValue)): Result (List (Pair String JValue)) Error =
+    def cmpKeysOnly (Pair k1 _) (Pair k2 _) =
+        k1 <~ k2
+    def simplifyKeyGroups = match _
+        Nil =
+            failWithError "groupBy invariant reached in deduplicateJObjectKeys"
+        (Pair key value), ps =
+            (value, map getPairSecond ps)
+            | rfoldr simplifyValues JNull
+            | rmap (Pair key)
+    def simplifyValues: JValue => JValue => Result JValue Error = match _ _
+        val JNull =
+            Pass val
+        JNull acc =
+            Pass acc
+        (JString val) (JString acc) if val ==~ acc =
+            Pass (JString acc)
+        (JInteger val) (JInteger acc) if val == acc =
+            Pass (JInteger acc)
+        (JDouble val) (JDouble acc) if val ==. acc =
+            Pass (JDouble acc)
+        (JBoolean val) (JBoolean acc) if (val && acc) || (!val && !acc) =
+            Pass (JBoolean acc)
+        (JArray val) (JArray acc) =
+            def pairEqual (Pair v1 v2) =
+                v1 ==/ v2
+            if len val == len acc && forall pairEqual (zip val acc) then
+                Pass (JArray acc)
+            else
+                Pass (JArray (val ++ acc))
```

Comment:
I think you're overthinking this one. Your rule is too magical. It can switch behaviours for a user-inexplicable reason. Pick one behaviour, only. I'd just go for `JArray (val ++ acc)` unconditionally, since that's the more helpful of the two options.